### PR TITLE
Add fastsimcoal 

### PR DESCRIPTION
**Pull request by ckyriazis**
Here is what I have so far for the dadi/fastsimcoal pipeline. I broke up the existing dadi pipeline to follow the format of the n_t pipeline, but other than breaking apart the simulation/sfs steps and changing the directory structure, I have left the dadi rules largely as-is. To that, I added rules to make an fsc-style SFS from the dadi SFS, and then set up and run a fsc analysis. 

I've been able to get this pipeline to run and produce output for dadi and fsc, but still have to deal with a number of issues (also summarized in the readme): 

1. dadi doesnt use the [0,0] entry of the 2D SFS, which fsc requires. The dadi sfs masks this entry, but unmasking it just produces a value of '0', so may need to go back into the ts_to_dadi_sfs function and  compute this value from the tree sequences. 

2. The IM model that fsc is running is more or less just the example IM model from fsc, so some time needs to be spent making sure this is as identical as possible to the IM model run by dadi. May also want to add other models for fsc to run. 

3. Figure out the best way to run fastsimcoal in the pipeline when downloading different versions for running on Mac/Windows/Linux (see here: http://cmpg.unibe.ch/software/fastsimcoal2/)

4. Add a plotting step to summarize and compare the parameter estimates from fsc and dadi. 

5. Set up the snakemake pipeline to run with a config file. 

6. QC, testing, and documentation. So far I've run this just with the human model but it would be good to run it with the drosophila two pop model. 

Finally, its worth mentioning that I havent used fastsimcoal or dadi before working on this so there may be things I've done incorrectly or overlooked. When the time comes to QC this pipeline, there are some people in the Lohmueller lab with more experience with this stuff that I can ask to look over it. I'll be able work on this during most of next week, but will actually be out for Memorial day weekend starting tomorrow without computer access.

### CODE REVIEW

**Comment by andrewkern** [comment]
awesome @ckyriazis. i'm going to pink @jgallowa07 for help on the review and eventual merge

**Comment by ckyriazis** [comment]
Just a quick update here - I think I've solved the issue with #1 above, I'll push an update early next week.

**Comment by RyanGutenkunst** [comment]
Pulling in @tjstruck for comments, since he wrote the dadi code

**Comment by jgallowa07** [comment]
re: "As of now, this snakemake pipeline only works with the Mac version of fastsimcoal2."

is this still true?

**Comment by ckyriazis** [comment]
Only in that the path to fastsimcoal in the shell command looks for the Mac
directory name - it would be easy to change that but I wasn’t sure the best
way to make that work with whatever version the user downloads.

**Comment by jgallowa07** [comment]
> Only in that the path to fastsimcoal in the shell command looks for the Mac directory name - it would be easy to change that but I wasn’t sure the best way to make that work with whatever version the user downloads.

Gotcha (My vote is linux if we have to hard-code/choose a default executable)

On a similar note, I'm having trouble running the whole pipeline, it seems to be failing in `rule run_fsc` First, I had to do a `chmod +x` to make `fsc26_linux64/fsc26` executable, and now it seems to be unable to open `IM.tpl` and `IM.est`: 
```jgallowa@sesame:~/popsim/analysis/two_population_analysis$ fsc26_linux64/fsc26 fsc26 -t IM.tpl -n 100000 -d -e IM.est -M -L 40 -q
fastsimcoal was invoked with the following command line arguments:
fsc26_linux64/fsc26 fsc26 -t IM.tpl -n 100000 -d -e IM.est -M -L 40 -q 
Unable to open input file (IM.tpl). Exiting program.
Unable to open input file (IM.est)```

any thoughts?

**Comment by jgallowa07** [comment]
As for the pipeline overall, the dependencies and structure look great!

>   Figure out the best way to run fastsimcoal in the pipeline when downloading different versions for running on Mac/Windows/Linux (see here: http://cmpg.unibe.ch/software/fastsimcoal2/)

I personally think we should just stick with linux - If you really want both, you can tell the user to download the appropriate executable and simply rename the `fsc26_<dist>` directory to something like `fsc` so that the Snakefile doesn't need to be altered. 

 >   Add a plotting step to summarize and compare the parameter estimates from fsc and dadi.

I like the Idea of boxplots as a comparison.

>   Set up the snakemake pipeline to run with a config file.

This is super easy to do and definitely helps keep the repo clean since you're not constantly changing the Snakefile. 

>  QC, testing, and documentation. So far I've run this just with the human model but it would be good to run it with the drosophila two pop model.

As far as documentation, I would be slightly more clear on what needs to be done for fsc to be run.

**Comment by ckyriazis** [comment]
Are the IM.* files being copied into the directories with the SFSs? For fsc to run, you need the .tpl, .est, and .obs (which is the SFS) file all in the same place.

File: two_population_analysis/Snakefile [hunk 1]
```diff
@@ -1,73 +1,172 @@
...
+
+    threads: 2
+    run:
+        inputs = expand(output_dir + "/{seeds}/{chrms}_dadi_joint_sfs.fs", chrms=chrm_list,seeds=wildcards.seeds)
+        dadi_out_path = output_dir + "/" + wildcards.seeds + "/all_chrm_dadi_joint_sfs.fs"
+        fsc_out_path  = output_dir + "/" + wildcards.seeds + "/fsc_analysis/IM_jointDAFpop1_0.obs"
+        fastsimcoal.dadi_to_fsc_sfs(inputs, dadi_out_path, fsc_out_path, sample_size=num_sampled_genomes_per_replicate)
+
+
+rule fsc_setup:
```
Comment (jgallowa07) [inline two_population_analysis/Snakefile]: There is no dependency on this rule (meaning it never gets run)

File: two_population_analysis/Snakefile [hunk 2]
```diff
@@ -1,73 +1,172 @@
...
+    input: expand(output_dir + "/{seeds}/fsc_analysis/IM_jointDAFpop1_0.obs", seeds=seed_array)
+
+    output: output_dir + "/{seeds}/fsc_analysis/IM.{file_type}"
```
Comment (jgallowa07) [inline two_population_analysis/Snakefile]: "file_type" wildcard is not defined. I would replace this line with these two:
```
162         output_dir + "/{seeds}/fsc_analysis/IM.est",
163         output_dir + "/{seeds}/fsc_analysis/IM.tpl"
```
or something equivilent

File: two_population_analysis/Snakefile [hunk 3]
```diff
@@ -1,73 +1,172 @@
...
+
+    shell: "cp fsc_files/IM.est {output_dir}/{wildcards.seeds}/fsc_analysis/IM.est && \
+            cp fsc_files/IM.tpl {output_dir}/{wildcards.seeds}/fsc_analysis/IM.tpl"
+
 
+rule run_fsc:
```
Comment (jgallowa07) [inline two_population_analysis/Snakefile]: Another input to this should be `rules.fsc_setup.output`

File: two_population_analysis/Snakefile (hunk 1 above)
Comment (ckyriazis) [reply two_population_analysis/Snakefile]: Thanks - I noticed that today and fixed it. Will send updates by the end of the day.

**Comment by ckyriazis** [comment]
I just sent some updates to the add_fsc branch that I think should resolve the things @jgallowa07  caught above. They should also resolve the issue with the [0,0] entry in the SFS (#1 above) and improving the documentation for downloading and installing fastsimcoal. I tried to make the instructions for installation general for any platform - would be grateful if others could test this to see if they have any further issues. 

Right now I'm focusing on trying to make sure the IM model implementation in fsc and dadi is the same, and then plan to work on the plotting step to compare the results. Will keep you posted on any updates.

**Comment by tjstruck** [comment]
@ckyriazis just to let you know, dadi has been updated so that it can work with scipy 1.3.0.

The readme instructions worked for getting fastsimcoal running on a mac, linux, and windows. However, I've only tested the Snakefile on mac and linux.

I can run everything on Mac, but I did have to modify the path for it to locate the fsc26 folder (not sure if we want to do anything more generalized for fsc26's path).

On the linux machine (which is the HPC our lab uses) it's running into an issue with the sed command:
(base) -bash-4.1$ sed -i '' 's/SAMPLE_SIZE/20/g' homo_sapiens/1845187043/fsc_analysis/IM.tpl
sed: can't read s/SAMPLE_SIZE/20/g: No such file or directory

**Comment by ckyriazis** [comment]
Thanks for the feedback @tjstruck. I'll change the readme regarding the issue with scipy 1.3. Also, I did have the instruction in the readme to change the name of the fsc directory (mv fsc26_* fsc26), but I'm sure thats easy for a user to miss. Any thoughts on a better way to do this?

I tried making the sed command portable on mac and linux following these guidelines: https://stackoverflow.com/questions/16745988/sed-command-with-i-option-in-place-editing-works-fine-on-ubuntu-but-not-mac

So the new command would be: sed -i.bak 's/SAMPLE_SIZE/{num_sampled_genomes_per_replicate}/' {output_dir}/{wildcards.seeds}/fsc_analysis/IM.tpl && rm {output_dir}/{wildcards.seeds}/fsc_analysis/IM.tpl.bak"

I'll push an update to my branch with that soon but in the meantime feel free to paste that in and see if it works. 

Finally, on a different topic, would you be able to help code up a dadi IM model with no exponential growth and two migration parameters (so essentially a hybrid between the IM and split_mig model that are already included)? The fastsimcoal IM model has two migration parameters and as far as I know wont work with just one, and it would be ideal to have the models be the same. Let me know if this makes sense.

**Comment by tjstruck** [comment]
@ckyriazis, it looks like the new sed command works on both mac and linux.

I can update dadi's Demographics2D.py with a new model I'll add as IM_fsc, but just to make sure I'm understanding what we want:
The new model would be a modification of the split_mig model with a migration rate for each population (m12 and m21 like in dadi's IM model)?

The IM model includes a parameter "s" that represents the initial size of population 1 as a percentage of the ancestral population (with 1-s as the initial size of population 2). If I'm interpreting the fsc_files correctly, there isn't something similar to the "s" parameter in the IM model fsc uses. It just has the final population sizes, right?

**Comment by ckyriazis** [comment]
@tjstruck, yes thats correct - the same as the split_mig model but just with one additional m parameter, and no s parameter as in the dadi IM model. Thanks!

**Comment by tjstruck** [comment]
@ckyriazis, IM_fsc has been added to Demographics2D.py and pushed to the dadi repo. Let me know if you find an error.

**Comment by ckyriazis** [comment]
Thanks @tjstruck it appears to work! Just one question - what is the direction of the respective m parameters (i.e. pop 1 -> 2 vs. pop 2 -> 1)?

**Comment by tjstruck** [comment]
@ckyriazis 
m12 is migration from pop 2 -> 1
m21 is migration from pop 1 -> 2

**Comment by jgallowa07** [comment]
Would you mind squashing the commits for us, @ckyriazis?

**Comment by ckyriazis** [comment]
Hey @jgallowa07 , I just tried to squash using this as a guide: https://medium.com/@vanshbadkul/squashing-commits-in-git-a-step-by-step-guide-884581224adf
but all of the commits are still there. Any idea what I might be doing wrong?

**Comment by ckyriazis** [comment]
Also @tjstruck - could I ask you to code up an additional dadi model for use in the pipeline? What I'm trying to do now is fit a two population Gutenkunst model to the simulations for Africa and Europe to see how well it recovers the true parameters under the correct model. So the parameters would be: 

Nanc - size of Africa population before increase to modern size
Naf - size of africa population
Taf - time of increase from Nanc to Naf 
Nb - size of bottleneck out of africa population
Tb - timing of out of africa split
Neu0 - initial size of Europe population
Teu - timing of onset of exponential growth
Neu_final - final size of Europe population after exponential growth
m_af-b - symmetric migration rate between african and bottleneck population
m_af-eu - symmetric migration rate between african and europe population

So basically just the Gutenkust model ignoring any parameters having to do with the Asian population. Let me know if this makes sense.

**Comment by jgallowa07** [comment]
> Hey @jgallowa07 , I just tried to squash using this as a guide: https://medium.com/@vanshbadkul/squashing-commits-in-git-a-step-by-step-guide-884581224adf
> but all of the commits are still there. Any idea what I might be doing wrong?

I would follow Jerome's instructions found here: https://github.com/popgensims/stdpopsim/blob/master/docs/development.rst

**Comment by tjstruck** [comment]
@ckyriazis, we played with something similar before as part of starting the two population analysis and getting useful code in the Snakefile, however ignoring the Asian population will make the parameters non-recoverable.

The more appropriate comparison would be marginalize the spectrum in the Gutenkunst model to be a 2D spectrum for the African and European populations and fitting that to the simulated date. However, that would be very slow for fitting.

**Comment by RyanGutenkunst** [comment]
@ckyriazis @tjstruck If the goal is to test the fitting process, then the easiest way is to change the msprime simulations temporarily to remove migration from the Asian population into Europe and Africa. Then one could fit a two-population Africa and Europe model and expect to recover the parameters quickly.

**Comment by ckyriazis** [comment]
These are all good points, and I think it would be helpful to discuss and clarify what our goals are here. But my understanding at the moment is that we arent necessarily trying to test the fitting process (it seems to be doing well with the simpler IM models) as much produce a rough comparison of how fsc and dadi perform in inferring the parameters of a model that is a better fit than the IM model, even if it doesnt get the parameters 100% correct. Also, @petrelharp mentioned on our last call that he may be able to help with thinking about what the expected parameter values should be under the models that we're trying to fit?

**Comment by ckyriazis** [comment]
Also @jgallowa07 - I cleaned up the commits using the hard reset described in that document, since things were getting pretty messy. Let me know if everything looks OK.

**Comment by RyanGutenkunst** [comment]
Okay... If the goal is just to increase complexity, then the model you proposed makes sense. @tjstruck, please do add it.

**Comment by tjstruck** [comment]
@ckyriazis do you explicitly want a bottleneck population/parameter rather than just having Europeans split from Africans? Currently in the Gutenkunst model the bottleneck population splits from the African population, and then the bottleneck population splits into European and Asian populations. So I'm not sure it is worth using for fitting.

**Comment by ckyriazis** [comment]
@tjstruck, I think maybe leave it in so we can compare the inferred Nb and Neu0 parameters to the true ones? Unless you think it will make fitting the model substantially more difficult.

**Comment by tjstruck** [comment]
@ckyriazis I think I was overthinking somethings. Since the model is pretty specific to what the project, let's have it be in dadi_utils.py instead of the in the dadi repo.

See attached file for the function OoA_2D_Af_Eu_func.
[OoA_2D_Af_Eu.txt](https://github.com/popgensims/analysis/files/3282826/OoA_2D_Af_Eu.txt)

**Comment by ckyriazis** [comment]
@tjstruck, I just tried a test run and got a "Dimensionality of phi and lengths of ns and xxs do not all agree." error. Heres how I set the function and parameters after pasting the function into dadi_utils.py: 

`if demo_model == 'Gute2pop':
		func = OoA_2D_Af_Eu_func
		params = [1.5, 1.5, 1.5, 1.5, 0.5, 0.5, 0.5, 0.5, 0.5]
		lower_bound = [0, 0, 0, 0, 0, 0, 0, 0, 0]
		upper_bound = [50, 50, 50, 50, 10, 10, 10, 10, 10]`

Any idea what might be going wrong?

**Comment by tjstruck** [comment]
Sorry about that @ckyriazis, minor mistake at the end of the function when making the spectrum. This one should work, but let me know if you run into any other issues:
[OoA_2D_Af_Eu.txt](https://github.com/popgensims/analysis/files/3282920/OoA_2D_Af_Eu.txt)

**Comment by ckyriazis** [comment]
No worries! And thanks 👍

**Comment by andrewkern** [comment]
hey @ckyriazis- just checking on the status here as I've lost track. Is this fully implemented now? should we do a review?

**Comment by ckyriazis** [comment]
Hey @andrewkern - I'm still working on a few things, including testing the inference under a two population Gutenkunst model, running the pipeline with the two population Drosophila model and IM inference (I didnt anticipate how long these simulations would take..), and then once I finish with that I plan to work on other stuff like adding the config file, cleaning up the code in some places, etc. We also need to think a bit more about what the ground truth parameters would be when doing inference with these different models so I can add those to the plotting functions. 

So overall, still a fair amount to be done but I think we're getting close to having a finished pipeline for review. Will be helpful to talk tomorrow.

**Comment by tjstruck** [comment]
@ckyriazis Just a heads up, I've actually been trying to get things working for Drosophila for awhile. I have some msprime simulations that took 5 to 6 days with 2 samples per-population. I opened an issue in stdpopsim to check if I'm doing anything wrong.

The main thing that slows things down is the recombination map, so if you don't have much better luck with running simulations, it might be best to have the framework for Drosophila without the recombination map?

**Comment by ckyriazis** [comment]
@tjstruck thanks for this input - I must have missed the issue you opened. I had been running a simulation for 20 samples and one chromosome under the two population model for over a week until I killed it today. @andrewkern seemed to think that there might be an issue with the recombination map, and said they'd try to look into that. 

Another thing I noticed when looking through stuff again just now is that the simulations.py module that is used both by the n_t and two population pipeline imports only `homo_sapiens` from `stdpopsim` and not the other species. Could this be causing any problems?

**Comment by jradrion** [comment]
I posted this reply to the issue raised by @tjstruck, but I'll post it here too. 

"Andy and I have been looking into this, and we think this is standard behavior and not a bug. I had previously used msprime to simulate the same dmel recombination maps under equilibrium for a different project, and I was experiencing run times of ~1hr for simulating a single chromosome (2R) with Ne =2.5e5. The human models take significantly longer to run as well if you bump Ne up > 100,000. It might be worth considering abandoning recombination maps for models with very large Ne, if we will need to do many simulation reps."

**Comment by ckyriazis** [comment]
Thanks @jradrion. Is everyone fine with running this without a recombination map?

**Comment by jradrion** [comment]
@ckyriazis @tjstruck
` ts=msp.simulate(
             sample_size=4,
             Ne=2.5e5,
             mutation_rate=2.8e-9,
             recombination_map=recomb_map)`
This example for Dmel chromosome 2R just took 85 mins to run on my machine. I'm going to test one rep of the full analysis pipeline on our cluster to see what sort of runtimes we might be looking at.

**Comment by andrewkern** [comment]
that's @jradrion. can we get similar runtime benchmarks for the demographic models in drosophila?

**Comment by RyanGutenkunst** [comment]
And how is the scaling with sample size? Four is pretty small.

**Comment by jradrion** [comment]
I'm currently running a test to simulate chr2R under the Li & Stephan model for samples sizes 2,4,8,16,and 32. I will update with runtimes when the jobs finish.

**Comment by ckyriazis** [comment]
Hey all - I just pushed my (hopefully) finalized changes for the two population pipeline for IM inference. I set it up to run with a config file and updated the readme to mirror the one in the n_t pipeline. It should work with both the human and drosophila two population models, but the drosophila one remains untested until we figure out what to do with the simulations. 

There are a few remaining loose ends including:
1. The ground truth values to be included for the plots when inferring from the Gutenkunst and Li & Stephan models under an IM model. Hoping @petrelharp can provide some insight here. 

2. I had some trouble automating the number of fsc runs so currently it is hard coded to do 10 dadi and fsc runs for each simulation replicate. Ideally I would like this to be something that can be changed in the config file, although it seems to converge fine for the IM model with just 10 runs. 

Let me know what you all think and if there are any changes that need to be made. I wont have much time to work on this over the next couple weeks, but will be able to focus on it again starting on July 8. Tagging @dortegadelv here so you can start working on adding SMC++ to this pipeline.

**Comment by petrelharp** [comment]
We've got a working method for this over at https://github.com/tskit-dev/msprime/pull/779 once this is merged.

I'm still working through how to plug this in to this code, @ckyriazis, but here's the basic idea: to compute mean coalescence time for two samples from today in population 0 of the Gutenkunst model:
```
import msprime
from stdpopsim import homo_sapiens

model = homo_sapiens.GutenkunstThreePopOutOfAfrica()
ddb = msprime.DemographyDebugger(**model.asdict())
m00 = ddb.mean_coalescence_time([0,0,2], rtol=.01)
print(m00)
>> 18094.08790235354
```
Divided by 2, this is a natural thing to compare to as the "true" Ne.

**Comment by petrelharp** [comment]
Ok, we've (finally) got `mean_coalescence_time()` merged into msprime:
  https://github.com/tskit-dev/msprime/commit/fc8ad0b5c8ff9c02211fb06bb1402e4fc98c5e49

**Comment by ckyriazis** [comment]
Thanks @petrelharp! I'll play around with this when I have time in a week or so.

**Comment by andrewkern** [comment]
@ckyriazis this PR might now be broken with the changes to `generation_time` that I just merged. would you mind double checking that?

**Comment by ckyriazis** [comment]
@andrewkern I just checked and I dont think the two population pipeline (as it currently is) needs the generation time specified anywhere.

**Comment by ckyriazis** [comment]
@petrelharp I'm a bit unclear on how to go from these mean coalescence times to a point estimate of the 'truth' to compare to the inferred values under an IM model. If I understand correctly, this method prints out mean coalescence time for a population at any given generation (though how do you specify the generation?) but we would need some sort of average over the duration of each epoch in the IM model we're trying to fit, right?

**Comment by ckyriazis** [comment]
Hey @petrelharp just following up here - could you help clarify how you think this method should be added to the pipeline?

**Comment by petrelharp** [comment]
Sorry, missed this.

And, I guess we hadn't thought this through sufficiently. If you have a method to infer Ne of a *single* population, then the natural thing to expect that it will infer is the mean time since coalescence between two genomes from the sample (so, we don't need to average over time or anything). But, of course, an IM model does something more complicated than this. Given all the mean pairwise coalescence times, we could work out the best-fitting IM model, although this is [complicated and maybe nonunique](https://www.biorxiv.org/content/10.1101/451328v3) and starts to be an inference method.

So, here's my proposal. We'd like to compare the inferred values to *something* because, well, that's the point? And when we're fitting a model that we know doesn't actually match the truth, what do we compare to? This is the situation we also have in real life, where we don't really believe the models; so we should compare the results to whatever it is about the real world that we'd like the methods to describe. And, maybe there's a few of these:
0. Census size, today. Why not?
1. Census size, averaged somehow over time. A harmonic mean over the time that the simulation is run for isn't perfect because it depends on how long you run the simulation for: the long-ago stuff shouldn't affect things. But, it's an easy thing to get, so a good comparison. An alternative would be a harmonic mean over time, weighted by probability of not having yet coalesced.
2. Mean coalescence time. Also, why not? It's a useful point of reference.

I think it's worth thinking ahead to how we will write up the results here. What exactly will we want to say about the results?

tldr; I'm proposing just adding the mean coalescence time within each pop to the plot, as well as with other census-size-derived points of reference.

**Comment by andrewkern** [comment]
> 0. Census size, today. Why not?

Sure. Could easily do this for N(t) too

>     1. Census size, averaged somehow over time. A harmonic mean over the time that the simulation is run for isn't perfect because it depends on how long you run the simulation for: the long-ago stuff shouldn't affect things. But, it's an easy thing to get, so a good comparison. An alternative would be a harmonic mean over time, weighted by probability of not having yet coalesced.
> 

Is there an exponential weighting to population size going back in time that would be the logical average to take here? Sort of like when people do exponential moving averages on times series data, but fixed for our single observation at time=0?

**Comment by RyanGutenkunst** [comment]
It's not a very sophisticated metric, but I would also include the simulated versus inferred divergence time. In my experience, that tends to be the "headline" number from these IM model inferences. (Even when there is ongoing gene flow that muddies the definition.)

**Comment by petrelharp** [comment]
> Is there an exponential weighting to population size going back in time that would be the logical average to take here?

You could weight by exp(-t/2Ne), but that'd just be a proxy for P(hasn't coalesced), which we can easily compute now...

**Comment by petrelharp** [comment]
>  also include the simulated versus inferred divergence time. In my experience, that tends to be the "headline" number

Oh, good point. Maybe we should really focus on this, where the comparison is clear.

**Comment by dortegadelv** [comment]
Hey @ckyriazis just to ask a quick question, how does your config.json file looks like? I want to reproduce your results while adding the smc++ results. I particularly want to know which "model" you are using in your config.json file.

{
    "seed" : 12345,
    "num_sampled_genomes_per_replicate" : 20,
    "replicates" : 10,
    "species" : "homo_sapiens",
    "model" : "GutenkunstThreePopOutOfAfrica",
    "genetic_map" : "HapmapII_GRCh37",
    "chrm_list" : "chr22",
    "generation_time" : 30
}

**Comment by ckyriazis** [comment]
Hey @dortegadelv, here is an example of one that I ran for testing: 

{
    "seed" : 12345,
    "num_sampled_genomes_per_replicate" : 20,  
    "replicates" : 2,
    "species" : "homo_sapiens",
    "model" : "GutenkunstThreePopOutOfAfrica",
    "genetic_map" : "HapmapII_GRCh37",
    "chrm_list" : "chr22",
}

So yours looks essentially the same, except for that I didnt need to specify a generation time for either fsc or dadi.

**Comment by ckyriazis** [comment]
Hey @petrelharp, I'm finally getting around to adding the mean_coalescence_time to the pipeline but am getting a: 
`AttributeError: 'DemographyDebugger' object has no attribute 'mean_coalescence_time'`
even after updating msprime to version 0.7.1 using conda. Should I be installing directly from the source?

**Comment by petrelharp** [comment]
Hm, that's right: the function was added in [this commit](https://github.com/tskit-dev/msprime/commit/485761207ba74be99cf24ee50680bfe881e42ec7), after the 0.7.1 release. You can get it by installing from source; maybe we'll need to push another release of msprime?

**Comment by jeromekelleher** [comment]
We can push out 0.7.2 easily enough: there's been a low-level bugfix for some extremely rare events, so arguably we should make this available ASAP anyway. Just let me know.

**Comment by ckyriazis** [comment]
Thanks @jeromekelleher, that would be helpful.

**Comment by jeromekelleher** [comment]
Msprime 0.7.2 has been released on PyPI @ckyriazis --- conda packages should follow in about a day.

**Comment by dortegadelv** [comment]
We can close this branch now. #39  now supercedes this branch. The pipeline from #39 now runs fastsimcoal, smc++ and dadi taking into account any demographic history and number of chromosomes.

**Comment by andrewkern** [comment]
thanks @dortegadelv. I will close this PR.
